### DISCUSSION TITLE
Room 타이머 구현

### DISCUSSION THREAD

**Discussion by waterricecake** [mafia-together/backend#28] [General]
제가 이번에 Timer 이동하는거 구현해봤는데 계속 고민되드만요. 일단 요약하자면 상태패턴 + 연쇄 패턴으로 방 상태를 구현하였는데 각 구현체가 Timer를 돌려서 다음 구현체를 생성하고 돌리는 형식으로 구현했습니다.
#27, 그리고 branch feature/27 보시면서 확인해주세요

- interface (상태패턴)
<img src="https://github.com/mafia-together/backend/assets/91263263/1782cfd3-2af2-41e5-84d0-4ee3534b06f2" width="400" height="200"/>

- 구현체별 호출 (연쇄패턴)
<img src="https://github.com/mafia-together/backend/assets/91263263/f15f4d8c-91ff-4d3c-b334-781cd6ea8b5d" width="600" height="200"/>

이게 구현하면서 TimerTask에서 run을 통해 setNext(Room room)을 호출하면서 room을 넣어줄려면 구현체가 room을 필드로 가지고 있어야 해서 어쩔수없이 state와 Room이 순환참조를 하게되는데 찝찝한데 그래서 한번 여러분의 생각을 들어보고 싶어서 올려봅니다.

참고로 상태패턴 + 연쇄패턴을 사용하게된 이유는 
상태의 행동을 (timer지나가는것) 상태가 관리하기 때문에 방의 역할이 줄어듬
상태 자체를 상태들을 생성하고 바꿔주는 역할도 상태가 하게됨
추후 각 상태별 정책을 추가하기 편할것이라고 판단 (run 혹은 startTimer 에 새로운 정책을 추가하면됨)

아 또 테스트 아직 고민 중이네요... timer 지나게 테스트 해보신분 있으신가요?

**Comment by waterricecake** [comment]
두번째로 고민해 본건데 우리가 /rooms/info 할때 어차피 startTime, endTime timestamp로 보내줘야하자나요? 그래서 room에 startTime, endTime 기록해두고 request /rooms/info 올때 requestheader의 timeStamp(header에는 항상 시간이 저장되니까)로 현재 시각 비교해서 그 시간 이 후이면 다음 state로 넘어가게 구현하는것...
이렇게 생각한 이유는 각 room별 타이머 돌리면 방마다 타이머 만들어줘서 부하가 너무 심할 것 같다는 생각이 들어서 입니다. 그래서 또 생각한게 반장 껄 기준으로 돌리면 어떨까도 생각해봤는데..
아니면 localDateTime.now()로 현재시간이랑 비교해서 돌려주는 방법도 있겠지요. 여튼 timer를 사용하지 않고 시작시간 끝시간을 계산해서 기록해두고 polling 올때마다 현재시간(서버의 시간 or requestheader Time)과 비교하여 endTime넘어갈시 state를 바꾸는 방법

**Reply by thdwoqor** [reply]
지금  `localDateTime.now()로 현재시간이랑 비교해서 돌려주는 방법도 있겠지요.` 해당 방식을 사용하고있는거같은데 좋은거같아요
TimerTask에서 run <- 이방식은 어떻게 하는건지 잘이해가안됨 ..
